File: LeetCode/java/498_Diagonal_Traverse.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Stack;

class Solution {
    private int[] check(int a, int b, boolean isUpDirection, int[][] mat) {
        int[] answer = new int[mat.length * mat[0].length];
        int idx = 0;

        int[] upDirX = {1, 0};
        int[] upDirY = {0, 1};

        int[] downDirX = {0, 1};
        int[] downDirY = {1, 0};

        Stack<int[]> stack = new Stack<>();
        Queue<int[]> q = new LinkedList<>();
        boolean[][] visit = new boolean[mat.length][mat[0].length];
        q.add(new int[]{a, b});
        visit[a][b] = true;

        while (!q.isEmpty()) {
            int qSize = q.size();

            for (int k = 0; k < qSize; k++) {
                int[] cur = q.poll();
                int x = cur[0];
                int y = cur[1];

                answer[idx++] = mat[x][y];

                for (int i = 0; i < 2; i++) {
                    int nx, ny;

                    if (isUpDirection) {
                        nx = x + upDirX[i];
                        ny = y + upDirY[i];
                    } else {
                        nx = x + downDirX[i];
                        ny = y + downDirY[i];
                    }

                    if (nx < 0 || ny < 0 || nx >= mat.length || ny >= mat[0].length) {
                        continue;
                    }

                    if (visit[nx][ny] == false) {
                        visit[nx][ny] = true;
                        stack.push(new int[] {nx, ny});
                    }
                }
            }

            while (!stack.isEmpty()) {
                q.add(stack.pop());
            }

            if (isUpDirection) {
                isUpDirection = false;
            } else {
                isUpDirection = true;
            }
        }

        return answer;
    }

    public int[] findDiagonalOrder(int[][] mat) {
        return check(0, 0, true, mat);
    }
}

File: LeetCode/cpp/498_Diagonal_Traverse.cpp
class Solution {
public:
	vector<int> check(int a, int b, bool isUpDirection, vector<vector<int>>& mat) {
		vector<int> answer;

		int upDirX[] = { 1, 0 };
		int upDirY[] = { 0, 1 };

		int downDirX[] = { 0, 1 };
		int downDirY[] = { 1, 0 };

		stack<pair<int, int>> st;
		queue<pair<int, int>> q;
		vector<vector<bool>> visit(mat.size(), vector<bool>(mat[0].size(), false));
		q.push({ a, b });
		visit[a][b] = true;

		while (!q.empty()) {
			int qSize = q.size();

			for (int k = 0; k < qSize; k++) {
				int x = q.front().first;
				int y = q.front().second;
				q.pop();

				answer.push_back(mat[x][y]);

				for (int i = 0; i < 2; i++) {
					int nx, ny;

					if (isUpDirection) {
						nx = x + upDirX[i];
						ny = y + upDirY[i];
					}
					else {
						nx = x + downDirX[i];
						ny = y + downDirY[i];
					}

					if (nx < 0 || ny < 0 || nx >= mat.size() || ny >= mat[0].size()) {
						continue;
					}

					if (visit[nx][ny] == false) {
						visit[nx][ny] = true;
						st.push({ nx ,ny });
					}
				}
			}

			while (!st.empty()) {
				q.push(st.top());
				st.pop();
			}

			if (isUpDirection) {
				isUpDirection = false;
			}
			else {
				isUpDirection = true;
			}
		}

		return answer;
	}

	vector<int> findDiagonalOrder(vector<vector<int>>& mat) {
		return check(0, 0, true, mat);
	}
};